trace/client/src/main/scala/org/http4s/otel4s/middleware/trace/client/ClientMiddlewareBuilder.scala

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2023 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+package otel4s.middleware
+package trace
+package client
+
+import cats.effect.Concurrent
+import cats.effect.MonadCancelThrow
+import cats.effect.Outcome
+import cats.effect.Resource
+import cats.syntax.applicative._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import fs2.Stream
+import org.http4s.client.Client
+import org.typelevel.otel4s.trace.SpanKind
+import org.typelevel.otel4s.trace.StatusCode
+import org.typelevel.otel4s.trace.TracerProvider
+
+/** Middleware builder for wrapping an http4s `Client` to add tracing.
+  *
+  * @see [[https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-client]]
+  */
+class ClientMiddlewareBuilder[F[_]: TracerProvider: Concurrent] private (
+    urlRedactor: UriRedactor,
+    spanDataProvider: SpanDataProvider,
+    urlTemplateClassifier: UriTemplateClassifier,
+    headersAllowedAsAttributes: HeadersAllowedAsAttributes,
+    shouldTrace: RequestPrelude => ShouldTrace,
```

Comment:
upon further thought, it's weird to me that this is a `Function1` and not a SAM trait. I don't know if changing that has to happen in this PR (though given all the other changes, it would seem weird for it to happen elsewhere)